server/src/main/java/org/elasticsearch/injection/Planner.java

```diff
@@ -0,0 +1,131 @@
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSet;
+
+/**
+ * <em>Evolution note</em>: the intent is to plan one domain/subsystem at a time.
+ */
+final class Planner {
+    private static final Logger logger = LogManager.getLogger(Planner.class);
+
+    final List<InjectionStep> plan;
+    final Map<Class<?>, InjectionSpec> specsByClass;
+    final Set<Class<?>> requiredTypes; // The injector's job is to ensure there is an instance of these; this is like the "root set"
+    final Set<Class<?>> allParameterTypes; // All the injectable types in all dependencies (recursively) of all required types
+    final Set<InjectionSpec> startedPlanning;
+    final Set<InjectionSpec> finishedPlanning;
+    final Set<Class<?>> alreadyProxied;
+
+    /**
+     * @param specsByClass an {@link InjectionSpec} indicating how each class should be injected
+     * @param requiredTypes the classes of which we need instances
+     * @param allParameterTypes the classes that appear as the type of any parameter of any constructor we might call
+     */
+    Planner(Map<Class<?>, InjectionSpec> specsByClass, Set<Class<?>> requiredTypes, Set<Class<?>> allParameterTypes) {
+        this.requiredTypes = requiredTypes;
+        this.plan = new ArrayList<>();
+        this.specsByClass = unmodifiableMap(specsByClass);
+        this.allParameterTypes = unmodifiableSet(allParameterTypes);
+        this.startedPlanning = new HashSet<>();
+        this.finishedPlanning = new HashSet<>();
+        this.alreadyProxied = new HashSet<>();
+    }
+
+    /**
+     * Intended to be called once.
+     * <p>
+     * Note that not all proxies are resolved once this plan has been executed.
+     * <p>
+     *
+     * <em>Evolution note</em>: in a world with multiple domains/subsystems,
+     * it will become necessary to defer proxy resolution until after other plans
+     * have been executed, because they could create additional objects that ought
+     * to be included in the proxies created by this plan.
+     *
+     * @return the {@link InjectionStep} objects listed in execution order.
+     */
+    List<InjectionStep> injectionPlan() {
+        for (Class<?> c : requiredTypes) {
+            planForClass(c, 0);
+        }
+        return plan;
+    }
+
+    /**
+     * Recursive procedure that determines what effect <code>requestedClass</code>
+     * should have on the plan under construction.
+     *
+     * @param depth is used just for indenting the logs
+     */
+    private void planForClass(Class<?> requestedClass, int depth) {
+        InjectionSpec spec = specsByClass.get(requestedClass);
+        if (spec == null) {
+            throw new IllegalStateException("Cannot instantiate " + requestedClass + ": no specification provided");
+        }
+        planForSpec(spec, depth);
+    }
+
+    private void planForSpec(InjectionSpec spec, int depth) {
+        String indent;
+        if (logger.isTraceEnabled()) {
```

Comment:
Could we move this logging-only code to somewhere else so it's not right at the top of the method? Maybe just have it inline in the `logger.trace()` call?

Reply:
You mean recalculate the indentation string each time it's used?

Reply:
I don't know how to use lambdas in trace statements. 😅  Do you have an example I can follow?

Reply:
Ok I've made a change. Let me know if you think this is better.